lib/vmapi-client.js

```diff
@@ -9,11 +9,120 @@
  */
 
 /*
- * vmapi-client.js
+ * This module contains the VmapiClient object which has functions for
+ * interaction with VMAPI.
+ *
  */
 
-var restify = require('restify');
+var restify = require('restify-clients');
+
+
+/*
+ * VMAPI has default values for some fields when the fields are not actually
+ * in the VM objects in Moray. For some of those fields such as:
+ *
+ *  autoboot: null
+ *  brand: null
+ *  firewall_enabled: false
+ *  nics: []
+ *  platform_build_stamp: null
+ *  resolvers: []
+ *  server_uuid: null
+ *  snapshots: []
+ *  state: null
+ *  tags: {}
+ *  zone_state: null
+ *  zonepath: null
+ *
+ * vmadm *also* always has these, so any time they're different we still want to
+ * do an update. The ones VMAPI always includes that are *not* always in vmadm
+ * objects are put into VMAPI_UNSET_FIELDS here so that we can detect their
+ * default values and not force an update when the vmadm object does not have
+ * these but the VMAPI object does and has the default value.
+ *
+ * If we did do that update, we'd be updating such a VM on every restart since
+ * VMAPI would always give us an object with these properties even if the
+ * correct object were in Moray.
+ *
+ * We have also excluded:
+ *
+ *  customer_metadata: {}
+ *  internal_metadata: {}
+ *
+ * since these are not in the COMPARISON_FIELDS above. If they're added we'll
```

Comment:
They're long-gone. Fixed the comment here with bc5b29120e8cf0a6981713b1924ec953e2f1aa4d, but this was basically replaced with self.comparisonFields on the VmAgent object. It used to be a static set of fields, but now it's built using VmWatcher.WATCHED_FIELDS which itself is now coming from the watchers.